variant-remapping-automation/get_custom_assembly.py

```diff
@@ -0,0 +1,210 @@
+        base, ext = os.path.splitext(self.assembly_report_path)
+        return base + '_custom' + ext
+
+    @property
+    def output_assembly_fasta_path(self):
+        base, ext = os.path.splitext(self.assembly_fasta_path)
+        return base + '_custom' + ext
+
+    @property
+    def assembly_directory(self):
+        return os.path.dirname(self.assembly_fasta_path)
+
+    @cached_property
+    def required_contigs(self):
+        raise NotImplementedError
+
+    @cached_property
+    def assembly_report_rows(self):
+        """Parse the assembly report and return each row as a dict."""
+        with open(self.assembly_report_path) as open_file:
+            # Parse the assembly report file to find the header then stop
+            for line in open_file:
+                if line.lower().startswith("# sequence-name") and "sequence-role" in line.lower():
+                    self.assembly_report_headers = line.strip().split('\t')
+                    break
+            reader = DictReader(open_file, fieldnames=self.assembly_report_headers, dialect=excel_tab)
+            return [record for record in reader]
+
+    @cached_property
+    def extended_report_rows(self):
+        """Provide the list of assembly report rows extended with additional ones if there are any."""
+        if self.genbank_contig_to_add:
+            extended_report_rows = copy(self.assembly_report_rows)
+            for contig_dict in self.genbank_contig_to_add:
+                row = {
+                    "# Sequence-Name": contig_dict['genbank'],
+                    "Sequence-Role": "scaffold",
+                    "GenBank-Accn": contig_dict['genbank'],
+                    'Relationship': '<>'
+                }
+                if 'refseq' in contig_dict:
+                    row['RefSeq-Accn'] = contig_dict['refseq']
+                    row['Relationship'] = '='
+                extended_report_rows.append(row)
+        else:
+            extended_report_rows = self.assembly_report_rows
+        return extended_report_rows
+
+    @cached_property
+    def genbank_contig_to_add(self):
+        genbank_contigs = set(row['GenBank-Accn'] for row in self.assembly_report_rows)
+        return [contig_dict for contig_dict in self.required_contigs if contig_dict['genbank'] not in genbank_contigs]
+
+    @staticmethod
+    def get_written_contigs(fasta_path):
+        written_contigs = []
+        match = re.compile(r'>(.*?)\s')
+        if os.path.isfile(fasta_path):
+            with open(fasta_path, 'r') as file:
+                for line in file:
+                    written_contigs.extend(match.findall(line))
+        return written_contigs
+
+    @retry(tries=4, delay=2, backoff=1.2, jitter=(1, 3))
+    def download_contig_from_ncbi(self, contig_accession):
+        sequence_tmp_path = os.path.join(self.assembly_directory, contig_accession + '.fa')
+        self.info(contig_accession + " downloaded and added to FASTA sequence")
```

Comment:
This log line seems premature since it technically hasn't been downloaded yet...

Reply:
Not to mentioned unnecessary since there's another log couple of line bellow 😢. 
Removed